{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1>KEGG decoder examples and usage</h1>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    Designed to parse through a KEGG-Koala outputs (including blastKOALA, ghostKOALA, KOFAMSCAN) to determine the completeness of various metabolic pathways. KEGGDecoder was orginally designed by Ben Tull and is available at: https://github.com/bjtully/BioData/tree/master/KEGGDecoder.\n",
    "    This package is designed to easily visualize genome/MAGs metabolic capabilities through the outputs of KEGGDecoder.\n",
    "    \n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    It is based off of scripts and work done by Ben Tully, available at: https://github.com/bjtully/BioData/tree/master/KEGGDecoder and modified by Ben Tully, Roth Conrad, Taylor Reitner, Luiz Irber, and Jay Osvatic.\n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Possible outputs</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4 different outputs can be produced by this repository. By using the option --vizoption and 'static/interactive/tanglegram' you can pick the desired output.\n",
    "    \n",
    "    1.The FUNCTION_OUT.list generates a TSV version of the heat map. The first row contains pathway/process\n",
    "    names, subsequent rows contain submitted groups/genomes and fractional percentage of pathway/process\n",
    "\n",
    "    2. A 'static' figure output is an SVG file function_heatmap.svg. Each distinct identifier before the\n",
    "    underscore in the FASTA file will have a row\n",
    "\n",
    "    3. A 'interactive' figure output is an HTML file function_heatmap.html. Each distinct identifier before \n",
    "    the underscore in the FASTA file will have a row, but can be loaded into a browser and value will be displayed\n",
    "    by hovering over a cell with the mouse. Draw a box to zoom in on specific regions.\n",
    "\n",
    "    4.'tanglegram' -- under construction\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Input file generation and formatting for functional heatmap </strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The functional heatmaps are generated by the default output of BlastKOALA, GhostKOALA or KOFAMscan.\n",
    "\n",
    "Start with protein FASTA file (INPUT_PROTEIN.fasta). This file can be multiple genomes combined. Be sure your submitted FASTA file has headers that group genomes together, KEGG-decoder.py groups based on the name provided in FASTA header before the first underscore (_)\n",
    "\n",
    "Process protein sequences through KEGG-KOALA (GhostKoala, BlastKoala, or KOFAMSCAN) and download the tab-delimited KO assignment text file (KOALA_OUTPUT.txt)\n",
    "\n",
    "Your input should look like below!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Input format</strong>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "pycharm": {
     "is_executing": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "text": [
      "NORP96_1\r\nNORP96_2\r\nNORP96_3\r\nNORP96_4\r\nNORP96_5\r\nNORP96_6\tK04764\r\nNORP96_7\tK01890\r\nNORP96_8\tK01889\r\nNORP96_9\tK02887\r\nNORP96_10\tK02916\r\n"
     ],
     "output_type": "stream"
    }
   ],
   "source": [
    "!head Example_input_data.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Functional heatmap generation</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a static heatmap:\n",
    "    \n",
    "    !KEGG-decoder --input INPUT.txt --output OUTPUT.tsv --vizoption static\n",
    "       \n",
    "    This will generate an SVG called \"OUTPUT.svg\"\n",
    "        \n",
    "Here is our example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "pycharm": {
     "is_executing": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "text": [
      "/bin/sh: 1: KEGG-decoder: not found\r\n"
     ],
     "output_type": "stream"
    }
   ],
   "source": [
    "!KEGG-decoder --input Example_input_data.txt --output functions.tsv --vizoption static"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    You can look at this output as you would any SVG file. It is best viewed in your browser."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a interactive heatmap, with 4 clustering options (Euclidean, Correlation, In descending order of # \n",
    "genes annotated, In ascending order of # genes annotated):\n",
    "    \n",
    "        !KEGG-decoder --input INPUT.txt --output OUTPUT.tsv --vizoption interactive\n",
    "        \n",
    "        This will generate an html called \"OUTPUT.html\"\n",
    "        \n",
    "Here is our example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "pycharm": {
     "is_executing": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "text": [
      "/home/chris/venv_3/lib/python3.6/site-packages/KEGGDecoder/KEGG_clustering.py:36: FutureWarning: \r\n.ix is deprecated. Please use\r\n.loc for label based indexing or\r\n.iloc for positional indexing\r\n\r\nSee the documentation here:\r\nhttp://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\r\n  genome_df = genome_df.ix[sort_dex]\r\n/home/chris/venv_3/lib/python3.6/site-packages/KEGGDecoder/KEGG_clustering.py:42: FutureWarning: \r\n.ix is deprecated. Please use\r\n.loc for label based indexing or\r\n.iloc for positional indexing\r\n\r\nSee the documentation here:\r\nhttp://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\r\n  genome_df = genome_df.ix[sort_dex]\r\n"
     ],
     "output_type": "stream"
    }
   ],
   "source": [
    "!KEGG-decoder --input Example_input_data.txt --output functions.tsv --vizoption interactive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    This will create an interactive html with the clustering options in a pull down menu. It is also best veiwed \n",
    "    in your browser."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Input file generation and formatting for tanglegram </strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a tanglegram:\n",
    "\n",
    "\t!KEGG-decoder -i INPUT.tsv -o OUTPUT.tsv -v tanglegram --newick NEWICK-TREE --tangleopt NUM-ITERS\n",
    "\t\n",
    "Here is our example:"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "source": [],
    "metadata": {
     "collapsed": false
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}